### ISSUE TITLE
Add CrossEntropyLoss2d

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#551]
They exist in torch / THNN. Need to be auto-wrapped.
They are needed to implement FCNs for semantic segmentation and stuff.

https://github.com/torch/nn/blob/master/SpatialClassNLLCriterion.lua#L43-L50

**Comment by apaszke** [comment]
There is `nn.NLLLoss2d` and `nn.Softmax2d`. We only need to port `nn.LogSoftmax2d`.

**Comment by andreaskoepf** [comment]
To me these 2d modules are confusing e.g.
[SpatialSoftmax.py](https://github.com/pytorch/pytorch/blob/0f65c9267d5ec55584b0ec65acb5374c95af9c16/torch/legacy/nn/SpatialSoftMax.py) and [Softmax.py](https://github.com/pytorch/pytorch/blob/0f65c9267d5ec55584b0ec65acb5374c95af9c16/torch/legacy/nn/SoftMax.py) look identical. In the outer wrapper code the difference seems to be an argument check, e.g. input to the Softmax module has to be 2D and input to the Softmax2d module has to be 4d - but in the end they call the same function.

In `torch.nn.functional' there are no '2d' versions of these functions but you can simply call
`torch.nn.functional.log_softmax` or `torch.nn.functional.softmax` with 1d,2d,3d,4d tensors.

I took me a couple of minutes to understand this and I could imagine that it confuses also other users if they port a legacy model to the new function style and search for the corresponding 2d functions. Maybe it would be worth to add a hint in the documentation?

**Comment by apaszke** [comment]
Yeah, sure. We just need to add docs for functional interface first :confused:

**Comment by soumith** [comment]
this is no longer relevant
